File: project_euler/problem_56/powerful_digit_sum.py
# sum up individual digits of a number
def sumOfDigits(num):
    numsum = 0
    for digit in str(num):
        numsum += int(digit)
    return numsum


# return the max possible sum for a number of n digits(assume all digits are 9)
def maxPossibleSum(num):
    return 9*len(str(num))


maxSum = 0  # variable to hold the current largest digit sum
breakLoop = False

for a in range(99, 0, -1):
    for b in range(99, 0, -1):
        num = a ** b  # a^b
        digitSum = sumOfDigits(num)  # get the sum of the digits
        if digitSum > maxSum:
            maxSum = digitSum  # set a new max

        # counting down from 99, I can calculate the max possible digit sum(all digits are 9) and compare that to the
        # current max. If less, I can move on to next iteration
        maxPosSum = maxPossibleSum(num)
        if maxPosSum < maxSum:
            if a == b:  # if a and b are equal, I know i'm done
                breakLoop = True
                break
            else:
                break

    if breakLoop:
        break

File: project_euler/problem_56/PowerfulDigitSum.java
import java.math.BigInteger;

public class PowerfulDigitSum
{

    public static int sumOfDigits(BigInteger num)
    {
        int numsum = 0;
        for(char digit : num.toString().toCharArray())
        {
            numsum += Character.getNumericValue(digit);
        }
        return numsum;
    }

    public static int maxPossibleSum(BigInteger num)
    {
        return 9*num.toString().length();
    }

    public static int findMaxSum()
    {
        BigInteger num = new BigInteger("0");
        int digitSum = 0;
        int maxPosSum = 0;
        int maxSum = 0;

        for(int a = 99; a > 0; a--)
        {
            for(int b = 99; b > 0; b--)
            {
                BigInteger bigA = new BigInteger(Integer.toString(a));
                num = bigA.pow(b);
                digitSum = sumOfDigits(num);
                if(digitSum > maxSum)
                {
                    maxSum = digitSum;
                }

                maxPosSum = maxPossibleSum(num);
                if(maxPosSum < maxSum)
                {
                    if(a == b) { return maxSum; }
                    else { break; }
                }
            }
        }

        return maxSum;
    }

    public static void main(String[] args)
    {
        int result = findMaxSum();
        System.out.println(result);
    }
}
